torch_xla/test/metrics_test_wrapper.py

```diff
@@ -112,21 +152,72 @@ def _write_to_disk(output_string, output_filename):
-
-  sys.exit(sp_return_code)
+  # Retrieve any config files that affect this test.
+  # NOTE: these are ordered in increasing specificity. For example, if there
+  # was a base config that affects all tests and a specific config for a
+  # particular test, then the base config will be the first element in the
+  # list and the most specific config will be the last element.
+  ordered_config_dicts = []
+
+  path_to_search = FLAGS.test_folder_name
+  while True:
+    try:
+      f = gcsfs.open(os.path.join(FLAGS.root, path_to_search, 'config.json'))
+      ordered_config_dicts.insert(0, json.load(f))
+    except google.api_core.exceptions.NotFound:
+      pass
+    if path_to_search == '':
+      break
+    path_to_search = os.path.split(path_to_search)[0]
+  if not ordered_config_dicts:
+    raise ValueError('No config files found in {} or parent directories. '
+        'See example usage at top of metrics_test_wrapper.py'.format(
+            os.path.join(FLAGS.root, FLAGS.test_folder_name)))
+
+  # Consolidate configs into 1 dict by overwriting the least-specific configs
+  # with the increasingly more-specific configs.
+  config = ordered_config_dicts[0]
+  for c in ordered_config_dicts:
+    config.update(c)
+
+  # Collect historical metrics for this test and check for any regressions in
+  # the current run vs. the averages from previous runs.
+  metrics_storage_dir = os.path.join(
+      FLAGS.root, FLAGS.test_folder_name, _METRICS_HISTORY_DIR_NAME)
+  metrics_storage_dir += '/'
+  regression_test_config = config.get('regression_test_config', None)
+  if regression_test_config:
+    metrics_file_pattern = re.compile(_METRICS_FILE_PATTERN)
+    prev_metrics_files = [f for f in gcsfs.list(
+        metrics_storage_dir) if metrics_file_pattern.match(f.path)]
+    prev_metrics_strings = [gcsfs.open(
+        os.path.join(FLAGS.root, f.path), mode='rt').read() for f in
+        prev_metrics_files]
+    data_points = mcu.get_data_points_from_metrics_reports(
+        prev_metrics_strings)
+    regression_report = mcu.compare_metrics(
+        data_points, metrics, config=regression_test_config)
+  else:
+    print('Unable to check for metrics regressions. Config should contain '
+          '"regression_test_config" key -- see example at the top of '
+          'metrics_test_wrapper.py.', file=sys.stderr)
+    regression_report = ''
+
+  # Write the metrics from the current run to disk unless disabled by config.
+  if config.get('write_metrics_to_disk', True):
+    # Include the params for this invocation when saving metrics.
+    output_string = '{}\n\n{}'.format(FLAGS, metrics)
+    output_filename = os.path.join(
+        metrics_storage_dir, datetime.datetime.utcnow().strftime('%Y_%m_%d'))
+    _write_to_disk(output_string, output_filename)
+
+  if regression_report:
+    raise AssertionError(
+        'Non-empty XLA metrics regression report:\n{}'.format(
+            regression_report))
+  if sp_return_code:
```

Comment:
Shouldn't we bail out immediately upon process failure?

Reply:
I lean toward no but could go either way. Any immediate subprocess failure will result in no XLA metrics file, which leads to a fail fast Exception in `_find_correct_metrics_file`.

I think the most common case of non-zero subprocess return code will be failing the accuracy assertion, in which case we still want to save metrics and check for metrics regressions.

The bad case is when we fail midway through training since then we will save metrics for partial runs - but hopefully we will see the alert when the training job died during a partial run and we could clean up the partial metrics? 

Or maybe I could find a way to pass the number of trained epochs from the child process to this wrapper and only write/assert XLA metrics if we trained for all the desired epochs

Reply:
But if the process exits early with error, there is little point in comparing the metrics I think.
They might be totally off due to truncation.

Reply:
Yes exactly, so 3 cases:
1. process exits non-0 before training starts - **no current problem** since there's no metrics file
2. process exits non-0 after training finishes since accuracy too low - **no current problem** since desired behavior is to check XLA metrics and write them to disk
3. process exits non-0 during training - **this is a problem** since the metrics are not to be trusted.

Some solutions:
-  use one exit code for case 2 and a different one for case 3 in the child process
- parse stdout logs of child process to see if training finished or not or maybe to see if number of completed epochs == desired number of epochs
- add another field to XLA metrics like "finished training"
- add number of epochs to child process tensorboard summary and search for the summary in this wrapper to see if we completed all epochs

Reply:
Oh and another solution option:
Don't check or write metrics at all unless subprocess has exit code 0. This just means we lose metrics if accuracy regresses - which is hopefully rare